### PR TITLE
Serializer

### PR DESCRIPTION
**Pull request by ghost**
Write serialisable classes

### CODE REVIEW

File: Serializer/Serializer/NSArray+serializable.m [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    NSMutableString *res = [[NSMutableString alloc]init];
+    NSMutableString *append = [[NSMutableString alloc]init];
+    for(id obj in self){
+        if([obj respondsToSelector:@selector(serializeWithError:)]){
+            if((append = [obj serializeWithError:error]) == nil){
+                return nil;
+            }
+            NSRange range = NSMakeRange(-1, 0);
+            int location = 0;
+            while (range.location != NSNotFound){
```
Comment (Maxim-38RUS-Zabelin) [inline Serializer/Serializer/NSArray+serializable.m:23]: This cycle looks to me overcomplicated.

File: Serializer/Serializer/NSArray+serializable.m [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+                NSString *insert = [NSString stringWithFormat:@"NSArray[%d]/",[self indexOfObject:obj]];
+                [append insertString:insert atIndex:location+range.location+1];
+                location += range.location+[insert length];
+                range = [[append substringFromIndex:location] rangeOfString:@"\n"];
+            };
+            [res appendString:append];
+            [res appendString:@"\n"];
+        } else {
+            if (!!error) {
+                NSDictionary* userInfo = [[NSDictionary alloc] initWithObjectsAndKeys:
```
Comment (Maxim-38RUS-Zabelin) [inline Serializer/Serializer/NSArray+serializable.m]: Why won't you use the modern way to create a dictionary?

File: Serializer/Serializer/NSArray+serializable.m [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                                          [obj class],
+                                          @"Object of class is not serializable", nil];
```
Comment (Maxim-38RUS-Zabelin) [inline Serializer/Serializer/NSArray+serializable.m]: This doesn't look like a key...

File: Serializer/Serializer/NSDictionary+serializable.m [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    NSMutableString *res = [[NSMutableString alloc]init];
+    NSMutableString *append = [[NSMutableString alloc]init];
+    id obj = nil;
+    for(id key in self){
+        if(![key isKindOfClass:[NSString class]] &&
+           ![key isKindOfClass:[NSNumber class]]){
+            if (!!error) {
+                NSDictionary* userInfo = [[NSDictionary alloc] initWithObjectsAndKeys:
+                                          @"Wrong type of dictionary key",
+                                          [key class], nil];
```
Comment (Maxim-38RUS-Zabelin) [inline Serializer/Serializer/NSDictionary+serializable.m]: You can't really use a Class as a key.

File: Serializer/Serializer/NSDictionary+serializable.m [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+                location += range.location+[insert length];
+                range = [[append substringFromIndex:location] rangeOfString:@"\n"];
+            };
+            [res appendString:append];
+            [res appendString:@"\n"];
+        } else {
+            if (!!error) {
+                NSDictionary* userInfo = [[NSDictionary alloc] initWithObjectsAndKeys:
+                                          [obj class],
+                                          @"Object of class is not serializable", nil];
```
Comment (Maxim-38RUS-Zabelin) [inline Serializer/Serializer/NSDictionary+serializable.m]: The same mistake as above.

File: Serializer/Serializer/NSValue+serializable.m [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+//  Created by Admin on 25/07/14.
+//  Copyright (c) 2014 MyCompanyName. All rights reserved.
+//
+
+#import "NSValue+serializable.h"
+
+@implementation NSValue(serializable)
+-(NSMutableString*)serializeWithError:(NSError* __autoreleasing *)error{
+    NSMutableString *res = [[NSMutableString alloc]init];
+    if([(res=[[self description] mutableCopy]) rangeOfString:@"NSRect"].location == NSNotFound)
```
Comment (Maxim-38RUS-Zabelin) [inline Serializer/Serializer/NSValue+serializable.m]: Please use braces with any if clause. And also avoid writing such a long conditions.

File: Serializer/Serializer/NSArray+serializable.m (hunk 1 above)
Comment (ghost) [inline Serializer/Serializer/NSArray+serializable.m:23]: "for" cycle gets each object in array, serialisation of object is write in "append", then "while" cycle adds array name with index to each line in "append". what should I do with it?

Comment (Maxim-38RUS-Zabelin) [inline Serializer/Serializer/NSArray+serializable.m:23]: What for do you use the insertString:...? Why don't you use the appendString:...? I also don't understand what for you need a range arithmetic. That is why I told you the cycle looks overcomplicated.

File: Serializer/Serializer/Serializer.m [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+#import "Serializer.h"
+#import "NSDictionary+serializable.h"
+
+@implementation Serializer
+-(NSString*)serializeDictionary:(id)dictionary error:(NSError *__autoreleasing *)error{
+    if([dictionary isKindOfClass:[NSDictionary class]])
+        return [[dictionary serializeWithError:error] copy];
+    if (!!error) {
+        NSDictionary* userInfo = [[NSDictionary alloc] initWithObjectsAndKeys:
```
Comment (Maxim-38RUS-Zabelin) [inline Serializer/Serializer/Serializer.m]: You can't really use [dictionary class] as a key.

**Pull request closed without merging**